File: 2021/day15/main.jl
using DataStructures

Coord = Tuple{Int, Int}

# Input ------------------------
function parse_input(path)
    Dict(
        (i, j) => parse(Int, c)
        for (i, r) in enumerate(readlines(path)) 
            for (j, c) in enumerate(r)
    )
end

example = parse_input("day15/example.txt")
input = parse_input("day15/input.txt")

# Part 1 -----------------------
function get_neighbours(node, n_row, n_col)
    x, y = node
    ((x + δx, y + δy) 
        for (δx,δy)  in ((0, 1), (0, -1), (1, 0), (-1, 0)) 
            if 1 <= x + δx <= n_row && 1 <= y + δy <= n_col
    )
end

function dijkstra(cave)
    n_row = maximum(x for (x, _) in keys(cave))
    n_col = maximum(y for (_, y) in keys(cave))
    max_dist = sum(values(cave))
    dist = DefaultDict{Coord, Int}(max_dist)
    dist[(1,1)] = 0
    pq = PriorityQueue{Coord, Int}()
    pq[(1,1)] = 0
    while !isempty(pq)
        node = dequeue!(pq)
        for new_node in get_neighbours(node, n_col, n_row)
            new_dist = dist[node] + cave[new_node]
            dist[new_node] < new_dist && continue
            dist[new_node] = new_dist
            pq[new_node] = new_dist
        end
    end
    dist[(n_row, n_col)]
end

dijkstra(example)
dijkstra(input)

# Part 2 ----------------------------
function expand_cave(cave)
    n_row = maximum(x for (x, _) in keys(cave))
    n_col = maximum(y for (_, y) in keys(cave))
    new_cave = Dict{Coord, Int}()
    for (key, val) in cave, δx in 0:4, δy in 0:4
        x, y = key
        new_cave[(x + δx * n_row, y + δy * n_col)] = mod1(val + δx + δy, 9)
    end
    new_cave
end

example |> expand_cave |> dijkstra
input |> expand_cave |> dijkstra

File: 2021/day15/main.R
library(tidyverse)
library(tidygraph)

# Input --------------------------

parse_input <- function(file){
  file |>
    readLines(warn = FALSE) |>
    str_split("") |>
    map(as.integer) |>
    reduce(rbind) |>
    unname()
}

example <- parse_input(file.path("day15", "example.txt"))
input <- parse_input(file.path("day15", "input.txt"))


# Part 1 ------------------------
build_graph <- function(input){
  n_nodes <- length(input)
  n_row <- nrow(input)
  n_col <- ncol(input)
  nodes <- tibble(name = seq_len(n_nodes))

  col_offset <- rep(0:(n_col - 1) * n_col, each = n_col - 1)
  x <- rep(1:(n_row - 1), n_col)
  y <- seq_len(n_nodes - n_row)
  edges <-
    # vertical movements
    tibble(from = c(x, x + 1), to = c(x + 1, x)) |>
    mutate(across(c(from, to), `+`, col_offset)) |>
    # horizontal movements
    bind_rows(tibble(from = c(y, y + n_row), to = c(y + n_row, y))) |>
    mutate(weight = input[to])

  tbl_graph(nodes = nodes, edges = edges)
}

find_shortest_path <- function(input){
  input |>
    build_graph() |>
    to_shortest_path(1, length(input), weights = weight) |>
    pluck("shortest_path") %E>%
    mutate(total_dist = sum(weight)) |>
    slice(1) |>
    pull(total_dist)
}

testthat::expect_equal(example |> find_shortest_path(), 40)
input |> find_shortest_path()

# Part 2 ------------------------
expanded_cave <- function(cave){
  cave <- cbind(cave, cave + 1, cave + 2, cave + 3, cave + 4)
  cave <- rbind(cave, cave + 1, cave + 2, cave + 3, cave + 4)
  (cave - 1) %% 9 + 1
}

testthat::expect_equal(example |> expand_cave() |> find_shortest_path(), 315)
input |> expand_cave() |> find_shortest_path()
